### PR TITLE
Feature/ees 1250 upload zip endpoint

### PR DESCRIPTION
**Pull request by shakes90**
This pull request adds the functionality to allow a zipped datafile to be uploaded. At present the file entries will be validated in terms of containing the correct datafiles but will currently return as a failed upload until the importer functionality to stream the file contents is wired in.

### CODE REVIEW

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/FileUploadsValidatorService.cs [hunk 1]
```diff
@@ -60,72 +34,17 @@ public FileUploadsValidatorService(ISubjectService subjectService, IFileTypeServ
...
-            if (!IsCsvFile(metadataFilePath, metaFile))
-            {
-                return ValidationActionResult(MetaFileMustBeCsvFile);
-            }
-
-            if (await _subjectService.GetAsync(releaseId, name) != null)
-            {
-                return ValidationActionResult(SubjectTitleMustBeUnique);
-            }
+            return await ValidateDataFileNames(releaseId, dataFile.FileName, metaFile.FileName)
```
Comment (ntsim) [inline src/GovUk.Education.ExploreEducationStatistics.Admin/Services/FileUploadsValidatorService.cs:37]: Let's get unit tests in for all of these validations as they should be pretty easy to write.

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/FileUploadsValidatorService.cs [hunk 2]
```diff
@@ -195,13 +114,95 @@ private bool IsFileExisting(Guid releaseId, ReleaseFileTypes type, string name)
                 .Any(rf => String.Equals(rf.ReleaseFileReference.Filename, name, CurrentCultureIgnoreCase)
                 && rf.ReleaseFileReference.ReleaseFileType == type);
         }
+
+        private async Task<Either<ActionResult, bool>> ValidateDataFileNames(Guid releaseId, string dataFileName, string metaFileName)
```
Comment (ntsim) [inline src/GovUk.Education.ExploreEducationStatistics.Admin/Services/FileUploadsValidatorService.cs:118]: If the Either's right is a `Unit` you don't need to return a `true`.

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/FileUploadsValidatorService.cs [hunk 3]
```diff
@@ -195,13 +114,95 @@ private bool IsFileExisting(Guid releaseId, ReleaseFileTypes type, string name)
...
+
+            if (IsFileExisting(releaseId, ReleaseFileTypes.Metadata, metaFileName))
+            {
+                return ValidationActionResult(CannotOverwriteMetadataFile);
+            }
+
+            return true;
+        }
+
+        private async Task<Either<ActionResult, bool>> ValidateDataFileSizes(long dataFileLength, long metaFileLength)
```
Comment (ntsim) [inline src/GovUk.Education.ExploreEducationStatistics.Admin/Services/FileUploadsValidatorService.cs:165]: If the Either's right is a `Unit` you don't need to return a `true`.

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/FileUploadsValidatorService.cs [hunk 4]
```diff
@@ -195,13 +114,95 @@ private bool IsFileExisting(Guid releaseId, ReleaseFileTypes type, string name)
...
+            {
+                return ValidationActionResult(MetadataFileCannotBeEmpty);
+            }
+
+            return true;
+        }
         
-        public static readonly Dictionary<ReleaseFileTypes, IEnumerable<Regex>> AllowedMimeTypesByFileType = 
-            new Dictionary<ReleaseFileTypes, IEnumerable<Regex>>
+        private async Task<Either<ActionResult, bool>> ValidateDataFileTypes(Guid releaseId, IFormFile dataFile, IFormFile metaFile)
```
Comment (ntsim) [inline src/GovUk.Education.ExploreEducationStatistics.Admin/Services/FileUploadsValidatorService.cs:180]: If the Either's right is a `Unit` you don't need to return a `true`.

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/FileUploadsValidatorService.cs [hunk 5]
```diff
@@ -195,13 +114,95 @@ private bool IsFileExisting(Guid releaseId, ReleaseFileTypes type, string name)
...
+
+            if (!IsCsvFile(metadataFilePath, metaFile))
+            {
+                return ValidationActionResult(MetaFileMustBeCsvFile);
+            }
+
+            return true;
+        }
+        
+        private async Task<Either<ActionResult, bool>> ValidateSubjectName(Guid releaseId, string name)
```
Comment (ntsim) [inline src/GovUk.Education.ExploreEducationStatistics.Admin/Services/FileUploadsValidatorService.cs:198]: If the Either's right is a `Unit` you don't need to return a `true`.

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/FileUploadsValidatorService.cs [hunk 6]
```diff
@@ -195,13 +114,95 @@ private bool IsFileExisting(Guid releaseId, ReleaseFileTypes type, string name)
...
+
+        private async Task<Either<ActionResult, bool>> ValidateDataFileNames(Guid releaseId, string dataFileName, string metaFileName)
+        {
+            if (string.Equals(dataFileName.ToLower(), metaFileName.ToLower(), OrdinalIgnoreCase))
+            {
+                return ValidationActionResult(DataAndMetadataFilesCannotHaveTheSameName);
+            }
+            
+            if (dataFileName.IndexOf(" ", Ordinal) > -1 || 
+                dataFileName.IndexOfAny(Path.GetInvalidFileNameChars()) > -1)
```
Comment (ntsim) [inline src/GovUk.Education.ExploreEducationStatistics.Admin/Services/FileUploadsValidatorService.cs:126]: This is just a duplicate of the `FileContainsSpacesOrSpecialChars` directly below.

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/FileUploadsValidatorService.cs [hunk 7]
```diff
@@ -195,13 +114,95 @@ private bool IsFileExisting(Guid releaseId, ReleaseFileTypes type, string name)
...
+                return ValidationActionResult(DataFilenameCannotContainSpacesOrSpecialCharacters);
+            }
+            
+            if (FileContainsSpacesOrSpecialChars(dataFileName))
+            {
+                return ValidationActionResult(FilenameCannotContainSpacesOrSpecialCharacters);
+            }
+            
+            if (metaFileName.IndexOf(" ", Ordinal) > -1 || 
+                metaFileName.IndexOfAny(Path.GetInvalidFileNameChars()) > -1)
```
Comment (ntsim) [inline src/GovUk.Education.ExploreEducationStatistics.Admin/Services/FileUploadsValidatorService.cs:137]: This is just a duplicate of the `FileContainsSpacesOrSpecialChars` directly below.

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/FileUploadsValidatorService.cs [hunk 8]
```diff
@@ -178,11 +97,11 @@ private bool IsCsvFile(string filePath, IFormFile file)
...
         }
-        
-        private bool FileContainsSpacesOrSpecialChars(IFormFile file)
+
+        private bool FileContainsSpacesOrSpecialChars(string filename)
         {
-            return file.FileName.IndexOf(" ", Ordinal) > -1 || 
-                   file.FileName.IndexOfAny(Path.GetInvalidFileNameChars()) > -1;
+            return filename.IndexOf(" ", Ordinal) > -1 || 
+                   filename.IndexOfAny(Path.GetInvalidFileNameChars()) > -1;
```
Comment (ntsim) [inline src/GovUk.Education.ExploreEducationStatistics.Admin/Services/FileUploadsValidatorService.cs:104]: Don't think this validation is good enough. We ran into the problem just the other day that characters like `+` were allowed and this broke the 'Manage data' page completely for one of the releases.

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/ReleaseFilesService.cs [hunk 9]
```diff
@@ -90,12 +90,45 @@ public class ReleaseFilesService : IReleaseFilesService
...
-                            await _importService.Import(dataFile.FileName.ToLower(), metadataFile.FileName.ToLower(),
-                                    releaseId, dataFile);
+                            await _importService.Import(releaseId, dataFile.FileName.ToLower(), 
+                                metadataFile.FileName.ToLower(), dataFile, false);
                             return true;
                         });
                 });
         }
+        
+        public Task<Either<ActionResult, bool>> UploadDataFilesAsZipAsync(Guid releaseId,
```
Comment (ntsim) [inline src/GovUk.Education.ExploreEducationStatistics.Admin/Services/ReleaseFilesService.cs:100]: If the Either's right is a `Unit` you don't need to return a `true`.

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/DataZipArchiveService.cs [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+            {
+                { ReleaseFileTypes.DataZip, AllowedZipMimeTypes }
+            };
+        
+        public DataZipArchiveService(IFileTypeService fileTypeService)
+        {
+            _fileTypeService = fileTypeService;
+        }
+
+        public async Task<Either<ActionResult, Tuple<ZipArchiveEntry, ZipArchiveEntry>>> GetArchiveEntries(
```
Comment (ntsim) [inline src/GovUk.Education.ExploreEducationStatistics.Admin/Services/DataZipArchiveService.cs:34]: Let's add some unit tests for this.

**Review by ntsim — CHANGES_REQUESTED** [review]

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/FileUploadsValidatorService.cs (hunk 2 above)
Comment (shakes90) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Services/FileUploadsValidatorService.cs]: dlone

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/FileUploadsValidatorService.cs (hunk 3 above)
Comment (shakes90) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Services/FileUploadsValidatorService.cs]: done

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/FileUploadsValidatorService.cs (hunk 5 above)
Comment (shakes90) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Services/FileUploadsValidatorService.cs]: done

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/FileUploadsValidatorService.cs (hunk 6 above)
Comment (shakes90) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Services/FileUploadsValidatorService.cs]: yep - done

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/FileUploadsValidatorService.cs (hunk 7 above)
Comment (shakes90) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Services/FileUploadsValidatorService.cs]: yep done

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/FileUploadsValidatorService.cs (hunk 8 above)
Comment (shakes90) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Services/FileUploadsValidatorService.cs:105]: Add some extra val for ampersand - also on subject name/title which is what caused the initial problem.

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/FileUploadsValidatorService.cs (hunk 4 above)
Comment (shakes90) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Services/FileUploadsValidatorService.cs]: done

File: src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Services/DataArchiveValidationServiceTests.cs [hunk 11]
```diff
@@ -0,0 +1,7 @@
+namespace GovUk.Education.ExploreEducationStatistics.Admin.Tests.Services
+{
+    public class DataArchiveValidationServiceTests
+    {
+
```
Comment (ntsim) [inline src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Services/DataArchiveValidationServiceTests.cs:5]: Seems like we're missing tests for this?

**Review by ntsim — CHANGES_REQUESTED** [review]

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/FileUploadsValidatorService.cs (hunk 1 above)
Comment (shakes90) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Services/FileUploadsValidatorService.cs:38]: Have added unit tests just for validating the archive file is valid & invalid for now.

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/ReleaseFilesService.cs (hunk 9 above)
Comment (shakes90) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Services/ReleaseFilesService.cs]: done

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/DataZipArchiveService.cs (hunk 10 above)
Comment (shakes90) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Services/DataZipArchiveService.cs]: Have added unit tests just for validating the archive file is valid & invalid for now.

File: src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Services/DataArchiveValidationServiceTests.cs (hunk 11 above)
Comment (shakes90) [reply src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Services/DataArchiveValidationServiceTests.cs:39]: done

**Review by ntsim — APPROVED** [review]

**Pull request merged**